routing/http/types/record_announcement.go

```diff
@@ -0,0 +1,368 @@
+package types
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ipfs/boxo/util"
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/codec/dagjson"
+	"github.com/ipld/go-ipld-prime/datamodel"
+	"github.com/ipld/go-ipld-prime/node/basicnode"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/multiformats/go-multibase"
+)
+
+const SchemaAnnouncement = "announcement"
+const announcementSignaturePrefix = "PUT /routing/v1 announcement:"
+
+var _ Record = &AnnouncementRecord{}
+
+type AnnouncementScope string
+
+const (
+	AnnouncementBlock     AnnouncementScope = "block"
+	AnnouncementEntity    AnnouncementScope = "entity"
+	AnnouncementRecursive AnnouncementScope = "recursive"
+)
+
+// AnnouncementRequest represents a request to announce
+type AnnouncementRequest struct {
+	CID      cid.Cid
+	Scope    AnnouncementScope
+	TTL      time.Duration
+	Metadata []byte
+}
+
+// AnnouncementPayload is the [AnnouncementRecord] payload.
+type AnnouncementPayload struct {
+	CID       cid.Cid
+	Scope     AnnouncementScope
+	Timestamp time.Time
+	TTL       time.Duration
+	ID        *peer.ID
+	Addrs     []Multiaddr
+	Protocols []string
+	Metadata  string
+}
+
+// MarshalJSON marshals the [AnnouncementPayload] into a canonical DAG-JSON
+// representation of the payload.
+func (ap AnnouncementPayload) MarshalJSON() ([]byte, error) {
+	m := make(map[string]ipld.Node)
+	var err error
+
+	// TODO: or goes empty? Affects signature.
+	if ap.CID.Defined() {
+		m["CID"] = basicnode.NewString(ap.CID.String())
+	}
+
+	// TODO: is scope not set when empty? Affects signature.
+	if ap.Scope != "" {
+		m["Scope"] = basicnode.NewString(string(ap.Scope))
+	}
+
+	if !ap.Timestamp.IsZero() {
+		m["Timestamp"] = basicnode.NewString(util.FormatRFC3339(ap.Timestamp))
+	}
+
+	if ap.TTL != 0 {
+		m["TTL"] = basicnode.NewInt(ap.TTL.Milliseconds())
+	}
+
+	if ap.ID != nil {
+		m["ID"] = basicnode.NewString(ap.ID.String())
+	}
+
+	if ap.Metadata != "" {
+		m["Metadata"] = basicnode.NewString(ap.Metadata)
+	}
+
+	// TODO: or goes empty if len == 0? Affects signature.
+	addrs := []ipld.Node{}
+	for _, addr := range ap.Addrs {
+		addrs = append(addrs, basicnode.NewString(addr.String()))
+	}
+	m["Addrs"], err = makeIPLDList(addrs)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: or goes empty if len == 0? Affects signature.
+	protocols := []ipld.Node{}
+	for _, protocol := range ap.Protocols {
+		protocols = append(addrs, basicnode.NewString(protocol))
+	}
+	m["Protocols"], err = makeIPLDList(protocols)
+	if err != nil {
+		return nil, err
+	}
+
+	// Make final IPLD node.
+	nd, err := makeIPLDMap(m)
+	if err != nil {
+		return nil, err
+	}
+
+	// Encode it with the DAG-JSON encoder, which automatically sorts the fields
+	// in a deterministic manner.
+	var b bytes.Buffer
+	err = dagjson.Encode(nd, &b)
```

Comment:
@hacdias mind switching to DAG-CBOR to match https://github.com/ipfs/specs/pull/378?